pkg/util/kubernetes/autoscalers/datadogclient.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// NewDatadogFallbackClient generates a new client able to query metrics to a second Datadog endpoint if the first one fails
+func newDatadogFallbackClient(endpoints []config.Endpoint) (*datadogFallbackClient, error) {
+	if len(endpoints) == 0 {
+		return nil, log.Errorf("external_metrics_provider.endpoints must be non-empty")
+	}
+
+	defaultClient, err := newDatadogSingleClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ddFallbackClient := &datadogFallbackClient{
+		clients: []*datadogIndividualClient{
+			{
+				client:             defaultClient,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			},
+		},
+	}
+	for _, endpoint := range endpoints {
+		client := datadog.NewClient(endpoint.APIKey, endpoint.APPKey)
+		client.HttpClient.Transport = httputils.CreateHTTPTransport()
+		client.RetryTimeout = 3 * time.Second
+		client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+		client.SetBaseUrl(endpoint.URL)
+		ddFallbackClient.clients = append(
+			ddFallbackClient.clients,
+			&datadogIndividualClient{
+				client:             client,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			})
+	}
+
+	return ddFallbackClient, nil
+}
+
+func (ic *datadogIndividualClient) queryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	// TODO: check for err 500 (server side) instead of any kind of error because switching to DR doesn’t make sense if that’s the query which is bad.
+	if series, err := ic.client.QueryMetrics(from, to, query); err == nil {
+		ic.lastQuerySucceeded = true
+		ic.lastSuccess = time.Now()
+		ic.retryInterval /= 2
+		if ic.retryInterval < minRetryInterval {
+			ic.retryInterval = minRetryInterval
+		}
+		return series, err
+	} else { // revive:disable-line:indent-error-flow
+		ic.lastQuerySucceeded = false
+		ic.lastFailure = time.Now()
+		ic.retryInterval *= 2
+		if ic.retryInterval > maxRetryInterval {
+			ic.retryInterval = maxRetryInterval
+		}
+		return series, err
+	}
+}
+
+func (ic *datadogIndividualClient) hasFailedRecently() bool {
+	if ic.lastQuerySucceeded {
+		return false
+	}
+
+	return ic.lastFailure.Add(ic.retryInterval).After(time.Now())
+}
+
+func (cl *datadogFallbackClient) QueryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	errs := errors.New("Failed to query metrics on all endpoints")
+
+	skippedClients := []*datadogIndividualClient{}
+
+	for i, c := range cl.clients {
+
+		if c.hasFailedRecently() {
+			skippedClients = append(skippedClients, c)
+			continue
+		}
+
+		if series, err := c.queryMetrics(from, to, query); err == nil {
+			if i != cl.lastUsedClient {
+				log.Warnf("Switching external metrics source provider from %s to %s",
+					cl.clients[cl.lastUsedClient].client.GetBaseUrl(),
+					c.client.GetBaseUrl())
+			}
+			cl.lastUsedClient = i
+			return series, nil
+		} else { // revive:disable-line:indent-error-flow
+			log.Infof("Failed to query metrics on %s: %#v", c.client.GetBaseUrl(), err)
```

Comment:
Do you think it would make sense to add a note here that this isn't a hard error we will try another endpoint? Otherwise I think this might be a bit confusing when reading the logs.

Reply:
The log is already an `Info` which suggests it isn’t a fatal error.
Even if the DCA is able to fallback to another endpoint, the failure of a DC is serious enough to be logged.